modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/h2/affinity/PartitionExtractor.java

```diff
@@ -366,4 +379,171 @@ else if (rightParam != null)
         return new PartitionTableDescriptor(tbl.cacheName(), tbl.getName());
     }
+
+    /**
+     * Try to extract partitons from {@code op} assuming that it's between opoeration or simple range.
+     *
+     * @param op Sql operation.
+     * @return {@code PartitionSingleNode} if operation reduced to one partition,
+     *   {@code PartitionGroupNode} if operation reduced to multiple partitons or null if operation is neither
+     *   between nor simple range. Null also returns if it's not possible to extract partitions from given operation.
+     * @throws IgniteCheckedException If failed.
+     */
+    PartitionNode tryExtractBetween(GridSqlOperation op) throws IgniteCheckedException {
+        // Between operation (or similar range) should contain exact two children.
+        if (op.size() != 2)
+            return null;
+
+        if (op.child().size() != 2)
+            return null;
+
+        if (op.child(1).size() != 2)
+            return null;
+
+        if (op.child().child().size() != 0 || op.child(1).child().size() != 0)
+            return null;
+
+        // Try parse left AST.
+        GridSqlAst left = op.child();
+        GridSqlOperation leftOp;
+        GridSqlOperationType leftOpType;
+        GridSqlColumn leftCol;
+
+        if (left instanceof GridSqlOperation) {
+            leftOp = (GridSqlOperation)left;
+            leftOpType = leftOp.operationType();
+
+            if (leftOp.child() instanceof GridSqlColumn) {
+                leftCol = leftOp.child();
+
+                if (!(leftCol.column().getTable() instanceof GridH2Table))
+                    return null;
+            }
+            else
+                return null;
+        }
+        else
+            return null;
+
+        // Try parse right AST.
+        GridSqlAst right = op.child(1);
+        GridSqlOperation rightOp;
+        GridSqlOperationType rightOpType;
+        GridSqlColumn rightCol;
+
+        if (right instanceof GridSqlOperation) {
+            rightOp = (GridSqlOperation)right;
+            rightOpType = rightOp.operationType();
+
+            if (rightOp.child() instanceof GridSqlColumn) {
+                rightCol = rightOp.child();
+
+                if (!(rightCol.column().getTable() instanceof GridH2Table))
+                    return null;
+            }
+            else
+                return null;
+        }
+        else
+            return null;
+
+        // Check that both left and right AST use same column.
+        if (!leftCol.equals(rightCol))
+            return null;
+
+        // Check that columns might be used for partition prunning.
+        if (!((GridH2Table)leftCol.column().getTable()).isColumnForPartitionPruning(leftCol.column()))
+            return null;
+
+        // Check that left AST operation is '>' or '>=' and right AST operation is '<' or '<='.
+        if (!((leftOpType == GridSqlOperationType.BIGGER || leftOpType == GridSqlOperationType.BIGGER_EQUAL)
```

Comment:
Is case `rightOp == BIGGER && leftOp == SMALLER` is applicable for prunning? I suppose that right and left might be interchanged and still applicable for prunning.

Reply:
I believe it'll be part of next iteration of implementation. Now only range equal to between is supported.